### DISCUSSION TITLE
Add velocity weights or custom properties to resultant tile mesh

### DISCUSSION THREAD

**Discussion by niko-dellic** [isaac-mason/recast-navigation-js#431] [Q&A]
First of all, great work on recast-navigation-js, I'm super excited to see where this library goes.

I have three overlapping questions:

1. Can you assign weights to the tile mesh such that when an agent goes on that specific tile, they are faster? For example, a source geometry indicates bike lane versus a place to walk:

![image](https://github.com/user-attachments/assets/fbda62f1-27c3-46c4-a9e0-af4f2868460b)

2. Can this be considered when you are trying to compute the shortest path? (or in this case, fastest path)
3. More generally, can you assign other properties to the vertices of the resultant tiledNavMesh during its generation, such as assigning/transferring the UUID or object data of the origin geometry to the resultant navMesh?

![image](https://github.com/user-attachments/assets/df5a28e6-d0cf-4bbe-b2e9-fe68345d4615)

I'm currently generating these properties post hoc with the debugger mesh object, extracting the points, and raycasting downwards to get the origin object. This is very computationally intensive so it would be great if I could avoid it. 

Thanks again and looking forward to further development of recast-navigation-js!

Btw, I'm writing in pure vanilla so when I have a little free time, I would be happy to contribute some basic examples.

**Comment by isaac-mason** [comment]
Hi @niko-dellic!

This is possible, but requires some setup.

During the navmesh generation process you can mark areas with your own area ids.  e.g. see https://recast-navigation-js.isaacmason.com/?path=/story/advanced-custom-areas--compute-path
These could be anything you want, e.g. you could have an area id for "bike" and "walk"

If you want given areas to have different costs for traversal, e.g. corresponding to a faster/slower speed, you can configure a `dtQueryFilter` with different costs for traversing a polygon with the given area id. https://docs.recast-navigation-js.isaacmason.com/classes/index.QueryFilter.html

When moving your agents, either with a detour Crowd or your own implementation, you can get the polygon your agent is currently on, the get the area id you've given it with [getPolyArea](https://docs.recast-navigation-js.isaacmason.com/classes/index.RawModule.NavMesh.html#getPolyArea), and increase the speed of your agent.

**Reply by niko-dellic** [reply]
Hey @isaac-mason , thanks for the thorough answer! Ill take a stab at this in the coming week(s). Just to clarify, would configuring `dtQueryFilter ` then influence the shortest path via `navMeshQuery.computePath`?

Many thanks!

**Reply by isaac-mason** [reply]
Hey @niko-dellic, no worries.

You can either configure the NavMeshQuery default query filter `navMeshQuery.defaultFilter` or pass one to computePath, then computePath will consider the area costs defined.

Keen to hear how you go 🙂 

There aren't any good examples in the repo for all of this yet. If you feel inspired to add one please feel free!

**Reply by niko-dellic** [reply]
Will do!
